apollo-router/src/services/layers/apq.rs

```diff
@@ -32,99 +34,110 @@ struct PersistedQuery {
-                                .unwrap(),
-                            }];
-                            let res = RouterResponse::builder()
-                                .data(Value::default())
-                                .errors(errors)
-                                .context(req.context)
-                                .build()
-                                .expect("response is valid");
-
-                            Ok(ControlFlow::Break(res))
                         }
+                        (Some(apq_hash), _) => {
+                            if let Ok(cached_query) = cache.get(&apq_hash).await.get().await {
+                                let _ = req.context.insert("persisted_query_hit", true);
+                                tracing::trace!("apq: cache hit");
+                                req.originating_request.body_mut().query = Some(cached_query);
+                                Ok(ControlFlow::Continue(req))
+                            } else {
+                                tracing::trace!("apq: cache miss");
+                                let errors = vec![crate::error::Error {
+                                    message: "PersistedQueryNotFound".to_string(),
+                                    locations: Default::default(),
+                                    path: Default::default(),
+                                    extensions: serde_json_bytes::from_value(json!({
+                                          "code": "PERSISTED_QUERY_NOT_FOUND",
+                                          "exception": {
+                                          "stacktrace": [
+                                              "PersistedQueryNotFoundError: PersistedQueryNotFound",
+                                          ],
+                                      },
+                                    }))
+                                    .unwrap(),
+                                }];
+                                let res = RouterResponse::builder()
+                                    .data(Value::default())
+                                    .errors(errors)
+                                    .context(req.context)
+                                    .build()
+                                    .expect("response is valid");
+
+                                Ok(ControlFlow::Break(res))
+                            }
+                        }
+                        _ => Ok(ControlFlow::Continue(req)),
                     }
-                    _ => Ok(ControlFlow::Continue(req)),
-                }
+                })
+                    as BoxFuture<
+                        'static,
+                        Result<
+                            ControlFlow<<S as Service<RouterRequest>>::Response, RouterRequest>,
+                            BoxError,
+                        >,
+                    >
             },
-            service,
+            Buffer::new(service, DEFAULT_BUFFER_SIZE),
```

Comment:
🎉 this one makes sense!

Reply:
I'd like to remove it, but I have not found the way yet